aem-authoring-toolkit-plugin/src/main/java/com/exadel/aem/toolkit/core/handlers/container/common/ContainerHandler.java

```diff
@@ -0,0 +1,321 @@
+        // Note that "ignored tabs" setting is not inherited and is for current class only, unlike tabs collection
+        String[] ignoredTabs = componentClass.isAnnotationPresent(IgnoreTabs.class)
+            ? componentClass.getAnnotation(IgnoreTabs.class).value()
+            : new String[]{};
+
+
+        // Retrieve superclasses of the current class, from top of the hierarchy to the most immediate ancestor,
+        // populate tab registry and store fields that are within @Tab-marked nested classes
+        // (because we will not have access to them later)
+        Map<String, ContainerInfo> tabInstancesFromSuperClasses = getTabInstances(PluginReflectionUtility.getClassHierarchy(componentClass, false), annotation, containerName);
+
+        // Retrieve tabs of the current class same way
+        Map<String, ContainerInfo> tabInstancesFromCurrentClass = getTabInstances(Collections.singletonList(componentClass), annotation, containerName);
+
+        // Compose the "overall" registry of tabs.
+        Map<String, ContainerInfo> allTabInstances = getAllTabInstances(tabInstancesFromCurrentClass, tabInstancesFromSuperClasses);
+
+        // Get all *non-nested* fields from superclasses and the current class
+        List<Field> allFields = PluginReflectionUtility.getAllFields(componentClass);
+
+        // If tabs collection is empty and yet there are fields to be placed, fire an exception and create a default tab
+        if (allTabInstances.isEmpty() && !allFields.isEmpty()) {
+            PluginRuntime.context().getExceptionHandler().handle(new InvalidTabException(
+                exceptionMessage + componentClass.getSimpleName()
+            ));
+            if (containerName.equals(TABS)) {
+                allTabInstances.put(StringUtils.EMPTY, new ContainerInfo("newTab"));
+            }
+        }
+
+        // Render XML markup for all existing tabs
+        addTabs(allTabInstances, allFields, ignoredTabs, tabItemsElement, defaultTabName);
+
+        // Afterwards there still can be "orphaned" fields in the "all fields" collection. They are probably fields
+        // for which a non-existent tab was specified. Handle an InvalidTabException for each of them
+        handleInvalidTabException(allFields);
+    }
+
+    /**
+     * Method returning all properties stored in annotation as a map.
+     * @param annotation {@link Annotation} instance to analyze
+     * @return {@link Map<String, Object>}
+     */
+    public static Map<String, Object> getAnnotationFields(Annotation annotation) {
+        try {
+            Object handler = Proxy.getInvocationHandler(annotation);
+            Field f = handler.getClass().getDeclaredField("memberValues");
+            f.setAccessible(true);
+            return (Map<String, Object>) f.get(handler);
+        } catch (Exception ignored) {
+            //ignored
+        }
+        return new HashMap<>();
```

Comment:
Do we really need to return the new HashMap? We ignore exceptions here, but if an exception appears then unhandled NPE will be thrown (for example on the 266th line).

Reply:
Fixed